### PR TITLE
pass HTTP request args more explicitly to session

### LINKED ISSUE #5582: --num-procs X and curdoc().session_context.request.arguments don't go well together
If I add these two lines:
```
args = curdoc().session_context.request.arguments
print(args)
```
to `examples/app/sliders.py` and run bokeh with:
```
bokeh serve sliders.py --num-procs 2
```
and then reload the page it crashes in roughly 50% of the cases with:
```
Error running application handler <bokeh.application.handlers.script.ScriptHandler object at 0x7f468eafe978>: 'NoneType' object has no attribute 'arguments'
```
I guess one of the processes has the `request` not set properly. It works if I don't add `--num-procs`.

This happens with bokeh 0.12.3 and latest master (unrelated: master does not show the example at all, is it currently broken?)

For reference, this is the full example to reproduce:
```
''' Present an interactive function explorer with slider widgets.

Scrub the sliders to change the properties of the ``sin`` curve, or
type into the title text box to update the title of the plot.

Use the ``bokeh serve`` command to run the example by executing:

    bokeh serve sliders.py

at your command prompt. Then navigate to the URL

    http://localhost:5006/sliders

in your browser.

'''
import numpy as np

from bokeh.io import curdoc
from bokeh.layouts import row, widgetbox
from bokeh.models import ColumnDataSource
from bokeh.models.widgets import Slider, TextInput
from bokeh.plotting import figure

args = curdoc().session_context.request.arguments
print(args)

# Set up data
N = 200
x = np.linspace(0, 4*np.pi, N)
y = np.sin(x)
source = ColumnDataSource(data=dict(x=x, y=y))


# Set up plot
plot = figure(plot_height=400, plot_width=400, title="my sine wave",
              tools="crosshair,pan,reset,save,wheel_zoom",
              x_range=[0, 4*np.pi], y_range=[-2.5, 2.5])

plot.line('x', 'y', source=source, line_width=3, line_alpha=0.6)


# Set up widgets
text = TextInput(title="title", value='my sine wave')
offset = Slider(title="offset", value=0.0, start=-5.0, end=5.0, step=0.1)
amplitude = Slider(title="amplitude", value=1.0, start=-5.0, end=5.0)
phase = Slider(title="phase", value=0.0, start=0.0, end=2*np.pi)
freq = Slider(title="frequency", value=1.0, start=0.1, end=5.1)


# Set up callbacks
def update_title(attrname, old, new):
    plot.title.text = text.value

text.on_change('value', update_title)

def update_data(attrname, old, new):

    # Get the current slider values
    a = amplitude.value
    b = offset.value
    w = phase.value
    k = freq.value

    # Generate the new curve
    x = np.linspace(0, 4*np.pi, N)
    y = a*np.sin(k*x + w) + b

    source.data = dict(x=x, y=y)

for w in [offset, amplitude, phase, freq]:
    w.on_change('value', update_data)


# Set up layouts and add to document
inputs = widgetbox(text, offset, amplitude, phase, freq)

curdoc().add_root(row(inputs, plot, width=800))
curdoc().title = "Sliders"
```

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #5582
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by bryevdv** [comment]
@leopd @havocp thoughts? It turns out just copying `window.location.search.substr(1)` to the end of the websocket URL seems to work great. 

I also wrapped the `.requests` attr with a proxy that only exposes `.arguments`. This is because the actual request might be the original request or the later one for just the websocket, depending on whether the websocket happens to land on the original process or not. So I only wanted to expose info I could actually guarantee would be correct. If we can find a safe way to pass other things we can add them as needed. The proxy also removes some junk like `bokeh-session-id` etc from a a copy of the request arguments if needed.

**Comment by bryevdv** [comment]
To test I added the line:
```
print(curdoc().session_context.request.arguments)
```
to the sliders app run with `--num-procs=4`, then reloaded ~100 times with `?foo=10` and the output was always as expected.

**Comment by bryevdv** [comment]
Assuming folks agree this is acceptable I will add a migration note and some additional docs to clarify that only `request.arguments` is available.

**Comment by bryevdv** [comment]
OK I have updated the release notes I will merge when green.

File: bokeh/server/application_context.py [hunk 1]
```diff
@@ -15,6 +15,16 @@
 from bokeh.document import Document
 from bokeh.util.tornado import _CallbackGroup, yield_for_all_futures
 
+class _RequestProxy(object):
```
Comment (havocp) [inline bokeh/server/application_context.py]: I like this idea of limiting what's available to what's safe to use.

File: bokehjs/src/coffee/client.coffee [hunk 2]
```diff
@@ -134,6 +134,8 @@ class ClientConnection
 
     try
       versioned_url = "#{@url}?bokeh-protocol-version=1.0&bokeh-session-id=#{@id}"
+      if @args_string?.length > 0
+        versioned_url += "&#{@args_string}"
```
Comment (havocp) [inline bokehjs/src/coffee/client.coffee]: does this end up putting bokeh-protocol-version and bokeh-session-id in there twice or not? I'm not sure whether they could already be in the url

Comment (bryevdv) [inline bokehjs/src/coffee/client.coffee]: I don't believe so, but I will double check.

**Comment by bryevdv** [comment]
> The main downside vs. the "real" fix of making --num-procs have sticky sessions it that we do create the session twice, but that's not a regression.

Exactly. I wonder if there is a way to not create a session at all until the web socket connection is made? (And then just always create it in that instance) But that can be investigated later. 

Thanks for the quick look!

**Comment by bryevdv** [comment]
The fix for the unrelated CI problem is in master. This will be green when merged.

**Pull request merged**